linux/ibus-keyman/tests/km-dbus-test-server.c

```diff
@@ -0,0 +1,140 @@
+// DBus test server. The server will start and listen on a
+#include <gio/gio.h>
+#include <systemd/sd-bus.h>
+
+#ifndef KEYMAN_TEST_SERVICE_PATH
+#warning KEYMAN_TEST_SERVICE_PATH is undefined
+#define KEYMAN_TEST_SERVICE_PATH "."
+#endif
+
+#define KEYMAN_TESTSVC_BUS_NAME "com.keyman.ExitTestService"
+#define KEYMAN_TESTSVC_INTERFACE_NAME "com.keyman.ExitTestService.Exit"
+#define KEYMAN_TESTSVC_OBJECT_PATH "/com/keyman/ExitTestService/Exit"
+
+static GTestDBus *
+startup() {
+  GTestDBus *dbus = g_test_dbus_new(G_TEST_DBUS_NONE);
+
+  // Add the private directory with our in-tree service files.
+  g_test_dbus_add_service_dir(dbus, KEYMAN_TEST_SERVICE_PATH);
+
+  printf("Add service dir to %s\n", KEYMAN_TEST_SERVICE_PATH);
+
+  // Start the private D-Bus daemon
+  g_test_dbus_up(dbus);
+
+  const gchar *address = g_test_dbus_get_bus_address(dbus);
+
+  printf("Test server running on: %s\n", address);
+  FILE *f = fopen("/tmp/km-test-server.env", "w");
+  if (f) {
+    // write the address to a file which can be sourced
+    fprintf(f, "export DBUS_SESSION_BUS_ADDRESS=%s\n", address);
+    fclose(f);
+  }
+
+  return dbus;
+}
+
+static void
+finish(GTestDBus *dbus) {
+  // Stop the private D-Bus daemon
+  g_test_dbus_down(dbus);
+  g_object_unref(dbus);
+}
+
+static int32_t
+on_exit_method(sd_bus_message *msg, void *user_data, sd_bus_error *ret_error) {
+  g_debug("%s called", __FUNCTION__);
+  guint *exitFlag = (guint *)user_data;
+  *exitFlag       = TRUE;
+
+  return sd_bus_reply_method_return(msg, "");
+}
+
+static const sd_bus_vtable exit_test_service_vtable[] = {
+  SD_BUS_VTABLE_START(0),
+    SD_BUS_METHOD("Exit", "", "", on_exit_method, 0),
+  SD_BUS_VTABLE_END
+};
+
+static const void exit_loop(sd_bus_slot *slot, sd_bus *bus) {
+  sd_bus_release_name(bus, KEYMAN_TESTSVC_BUS_NAME);
+  sd_bus_slot_unref(slot);
+  sd_bus_close_unref(bus);
+}
+
+static void
+loop() {
+  sd_bus_slot *slot = NULL;
+  sd_bus *bus       = NULL;
+  int ret;
+
+  ret = sd_bus_open_user(&bus);
+
+  if (ret < 0) {
+    g_error("Failed to connect to system bus: %s", strerror(-ret));
+    return;
```

Comment:
Does this need `exit_loop(slot, bus);`?

Reply:
Not really, since opening the user bus failed `bus` is not set, nor `slot`, but for consistency it would be better to add it.

Reply:
This is because [C++ RAII](https://stackoverflow.com/questions/161177/does-c-support-finally-blocks-and-whats-this-raii-i-keep-hearing-about) means we don't get nice things (`try`/`finally`) -- otherwise we need to wrap these variables in a class for automatic destruction, which is tedious boilerplate. Another pattern I see used is `goto loop_cleanup`, which at least can keep the cleanup inside the same function, and is thus a little clearer, a little shorter, and avoids accidentally missing the `return` statement, or calling the cleanup twice (which in the current model would be _bad_ because it would result in double-frees):

```
  if (ret < 0) {
    g_error("Failed to connect to system bus: %s", strerror(-ret));
    goto cleanup;
  }

  // ... function continues ...

cleanup:
  if (bus)  sd_bus_release_name(bus, KEYMAN_TESTSVC_BUS_NAME);
  if (slot) sd_bus_slot_unref(slot);
  if (bus)  sd_bus_close_unref(bus);
}
```

Reply:
> I don't like gotos, so I put the whole thing in a class to benefit from C++ RAII 😄

I thought there was a good chance that you would say that! 🤣